File: sdk/nodejs-sdk/src/protocols/v1_0/CommittedAnswer.js
'use strict'
const utils = require('../../utils')
const Protocol = require('../Protocol')
/**
 * 1.0 CommittedAnswer protocol
 * The CommittedAnswer protocol allows one self-sovereign party ask another self-sovereign a question and get a answer.
 * The answer can be signed (not required) and must be one of a specified list of responses
 * @extends Protocol
 */
class CommittedAnswer extends Protocol {
  /**
   * Constructor for the 1.0 CommittedAnswer object. This constructor creates an object that is ready to ask
   * the given question
   *
   * @param forRelationship the relationship identifier (DID) for the pairwise relationship that will be used
   * @param threadId the thread id of the already started protocol
   * @param question The main text of the question (included in the message the Connect.Me user signs with their private key)
   * @param answerStr the given answer from the list of valid responses given in the question
   * @param descr Any additional information about the question
   * @param validResponses The given possible responses.
   * @param signatureRequired if a signature must be included with the answer
   * @return 1.0 CommittedAnswer object
   *
   * @property {String} msgFamily - 'committedanswer'
   * @property {String} msgFamilyVersion - '1.0'
   * @property {String} msgQualifier - 'Community Qualifier'
   * @property {String} this.msgNames.ASK_QUESTION - 'ask-question'
   * @property {String} this.msgNames.ANSWER_QUESTION - 'answer-question'
   * @property {String} this.msgNames.GET_STATUS - 'get-status'
   * @property {String} this.msgNames.ANSWER_GIVEN - 'answer-given'
   */
  constructor (forRelationship, threadId = null, question = null, answerStr = null, descr = null, validResponses = null, signatureRequired = null) {
    const msgFamily = 'committedanswer'
    const msgFamilyVersion = '1.0'
    const msgQualifier = utils.constants.COMMUNITY_MSG_QUALIFIER
    super(msgFamily, msgFamilyVersion, msgQualifier, threadId)
    this.forRelationship = forRelationship
    this.question = question
    this.answerStr = answerStr
    this.descr = descr
    this.validResponses = validResponses
    if (signatureRequired) {
      this.signatureRequired = signatureRequired
    } else {
      this.signatureRequired = true
    }

    this.msgNames.ASK_QUESTION = 'ask-question'
    this.msgNames.ANSWER_QUESTION = 'answer-question'
    this.msgNames.GET_STATUS = 'get-status'
    this.msgNames.ANSWER_GIVEN = 'answer-given'
  }

  /**
     * Creates the control message without packaging and sending it.
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the constructed message (JSON object)
     *
     * @see #ask
     */
  async askMsg (context) {
    var msg = this._getBaseMessage(this.msgNames.ASK_QUESTION)
    msg = this._addThread(msg)
    msg['~for_relationship'] = this.forRelationship
    msg.text = this.question
    msg.detail = this.descr
    msg.valid_responses = this.validResponses
    msg.signature_required = this.signatureRequired
    return msg
  }

  /**
     * Creates and packages message without sending it.
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the byte array ready for transport
     *
     * @see #ask
     */
  async askMsgPacked (context) {
    return this.getMessageBytes(context, await this.askMsg(context))
  }

  /**
     * Directs verity-application to ask the question
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     */
  async ask (context) {
    await this.sendMessage(context, await this.askMsgPacked(context))
  }

  /**
     * Creates the control message without packaging and sending it.
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the constructed message (JSON object)
     *
     * @see #answer
     */
  async answerMsg (context) {
    var msg = this._getBaseMessage(this.msgNames.ANSWER_QUESTION)
    msg = this._addThread(msg)
    msg['~for_relationship'] = this.forRelationship
    msg.response = this.answerStr
    return msg
  }

  /**
     * Creates and packages message without sending it.
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the byte array ready for transport
     *
     * @see #answer
     */
  async answerMsgPacked (context) {
    return this.getMessageBytes(context, await this.answerMsg(context))
  }

  /**
     * Directs verity-application to answer the question
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     */
  async answer (context) {
    await this.sendMessage(context, await this.answerMsgPacked(context))
  }

  /**
     * Creates the control message without packaging and sending it.
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the constructed message (JSON object)
     *
     * @see #status
     */
  async statusMsg (context) {
    var msg = this._getBaseMessage(this.msgNames.GET_STATUS)
    msg = this._addThread(msg)
    msg['~for_relationship'] = this.forRelationship
    return msg
  }

  /**
     * Creates and packages message without sending it.
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the byte array ready for transport
     *
     * @see #status
     */
  async statusMsgPacked (context) {
    return this.getMessageBytes(context, await this.statusMsg(context))
  }

  /**
     * Ask for status from the verity-application agent
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     */
  async status (context) {
    await this.sendMessage(context, await this.statusMsgPacked(context))
  }
}
module.exports = CommittedAnswer

File: sdk/python-sdk/verity_sdk/protocols/v1_0/CommittedAnswer.py
from typing import List

from verity_sdk.protocols.Protocol import Protocol
from verity_sdk.utils import COMMUNITY_MSG_QUALIFIER


class CommittedAnswer(Protocol):
    """
    The CommittedAnswer protocol allows one self-sovereign party ask another self-sovereign a question and get a
    answer. The answer can be signed (not required) and must be one of a specified list of responses
    """

    MSG_FAMILY = 'committedanswer'
    """the family name for the message family"""

    MSG_FAMILY_VERSION = '1.0'
    """the version for the message family"""

    # Messages
    ASK_QUESTION = 'ask-question'
    """Name for 'ask-question' control message"""
    GET_STATUS = 'get-status'
    """Name for 'get-status' control message"""
    ANSWER_QUESTION = 'answer-question'
    """Name for 'answer-question' signal message"""
    ANSWER_GIVEN = 'answer-given'
    """Name for 'answer-given' signal message"""

    # pylint:disable=too-many-arguments
    def __init__(self,
                 for_relationship: str,
                 thread_id: str = None,
                 question: str = None,
                 descr: str = '',
                 valid_responses: List[str] = None,
                 signature_required: bool = True,
                 answer_str: str = None,
                 msg_family: str = MSG_FAMILY,  # Used to override the msg_family for the
                 msg_family_version: str = MSG_FAMILY_VERSION,
                 msg_qualifier: str = COMMUNITY_MSG_QUALIFIER):
        """
        Args:
            for_relationship (str): the relationship identifier (DID) for the pairwise relationship that will be used
            thread_id (str): the thread id of the already started protocol
            question (str): The main text of the question (included in the message the Connect.Me user signs with their private key)
            descr (str): Any additional information about the question
            valid_responses (List[str]): The given possible responses.
            signature_required (bool): if a signature must be included with the answer
            answer_str (str): The given answer from the list of valid responses given in the question
        """
        super().__init__(msg_family, msg_family_version, msg_qualifier, thread_id)

        self.for_relationship: str = for_relationship
        self.question: str = question
        self.descr: str = descr
        self.valid_responses: List[str] = valid_responses
        self.signature_required: bool = signature_required
        self.answer_str: str = answer_str

    async def ask(self, context):
        """
        Creates the control message without packaging and sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
        """
        await self.send_message(context, await self.ask_msg_packed(context))

    def ask_msg(self, context):
        """
        Creates the control message without packaging and sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the constructed message (dict object)
        """
        msg = self._get_base_message(self.ASK_QUESTION)
        self._add_thread(msg)
        self._add_relationship(msg, self.for_relationship)
        msg['text'] = self.question
        msg['detail'] = self.descr
        msg['valid_responses'] = self.valid_responses or []
        msg['signature_required'] = self.signature_required
        return msg

    async def ask_msg_packed(self, context):
        """
        Creates and packages message without sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the bytes ready for transport
        """
        return await self.get_message_bytes(context, self.ask_msg(context))

    async def answer(self, context):
        """
        Directs verity-application to answer the question

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
        """
        await self.send_message(context, await self.answer_msg_packed(context))

    def answer_msg(self, context):
        """
        Creates the control message without packaging and sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the constructed message (dict object)
        """
        msg = self._get_base_message(self.ANSWER_QUESTION)
        self._add_thread(msg)
        self._add_relationship(msg, self.for_relationship)
        msg.response = self.answer_str
        return msg

    async def answer_msg_packed(self, context):
        """
        Creates and packages message without sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the bytes ready for transport
        """
        return await self.get_message_bytes(context, self.answer_msg(context))

    async def status(self, context):
        """
        Ask for status from the verity-application agent

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
        """
        await self.send_message(context, await self.status_msg_packed(context))

    def status_msg(self, context):
        """
        Creates the control message without packaging and sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the constructed message (dict object)
        """
        msg = self._get_base_message(self.GET_STATUS)
        self._add_thread(msg)
        self._add_relationship(msg, self.for_relationship)
        return msg

    async def status_msg_packed(self, context):
        """
        Creates and packages message without sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the bytes ready for transport
        """
        return await self.get_message_bytes(context, self.status_msg(context))
